javalib/src/main/scala/java/util/concurrent/ConcurrentSkipListSet.scala

```diff
@@ -0,0 +1,404 @@
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  override def add(e: E): Boolean =
+    if (e === null) throw new NullPointerException()
+    else inner.add(Box(e))
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = add(iter.next()) || changed
+    changed
+  }
+
+  override def clear(): Unit =
+    inner.clear() 
+
+  override def clone(): ConcurrentSkipListSet[E] =
+    new ConcurrentSkipListSet(this)
+
+  def comparator(): Comparator[E] = {
+    if (!provided) null
+    else  {
+      new Comparator[E] {
+        def compare(a: E, b: E) = ordering.compare(a, b)
+      }
+    }
+  }
+ 
+  override def contains(o: Any): Boolean =
+    inner.contains(Box(o.asInstanceOf[E]))
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toIndexedSeq.reverse.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  def descendingSet(): NavigableSet[E] = {
+    val descSetFun = () => {
+      val inverseOrdering = 
+        new Ordering[Box[E]] {
+         def compare(a: Box[E], b: Box[E]) = self.ordering.compare(b.inner,a.inner)
+        }
+      val retSet = new mutable.TreeSet[Box[E]]()(inverseOrdering)
+      retSet.addAll(inner)
+      retSet
+    }
+    new ConcurrentSkipListSet.NavigableView(this, descSetFun)
+  }
+
+  def first(): E = 
+    if (inner.isEmpty) throw new NoSuchElementException()
+    else inner.head.inner
+
+  def last(): E =
+    inner.lastOption.map(_.inner).getOrElse(null.asInstanceOf[E])
+
+  def ceiling(e: E): E =
+    tailSet(e, true).headOption.getOrElse(null.asInstanceOf[E])
+
+  def floor(e: E): E = 
+    headSet(e, true).lastOption.getOrElse(null.asInstanceOf[E])
+
+  def higher(e: E): E =
+    tailSet(e, false).headOption.getOrElse(null.asInstanceOf[E])
+    
+  def lower(e: E): E =
+    headSet(e, false).lastOption.getOrElse(null.asInstanceOf[E])
+
+  def pollFirst(): E = {
+    val polled = inner.headOption.map(_.inner)
+    if (polled.isDefined) {
+      val elem = polled.get
+      remove(elem)
+      elem
+    } else null.asInstanceOf[E]
+  }
+
+  def pollLast(): E = {
+    val polled = inner.lastOption.map(_.inner)
+    if (polled.isDefined) {
+      val elem = polled.get
+      remove(elem)
+      elem
+    } else null.asInstanceOf[E]
+  }
+
+  override def remove(o: Any): Boolean =
+    if (o === null) throw new NullPointerException()
+    else inner.remove(Box(o.asInstanceOf[E]))
+
+  override def removeAll(c: Collection[_]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = changed | inner.remove(iter.next)
```

Comment:
It's classical, I agree, but not with short-circuiting operators ^^ It's the combination of short-circuiting operator + foldLeft-style reduction that I fine obscure.

So in a sense you were quite right to use a non-short-circuiting `|`. But I happen to know their implementation in JS is a bit weird, so suboptimal. (Because there is no non-short-circuiting boolean or/and in JS, so we have to fake it.) I should have said this in the first place, rather than just saying "It's tricky.", my bad.